cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -567,11 +595,14 @@ private ProgramState getProgramState(boolean success, String failureInfo) {
    * Calls the destroy method of {@link ProgramLifecycle}.
    */
   private void destroy(final boolean succeeded, final String failureInfo, final ClassLoader mapReduceClassLoader) {
-    TransactionContext txContext = context.getTransactionContext();
+    final TransactionControl txControl = mapReduce instanceof ProgramLifecycle
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, MapReduce.class, mapReduce, "destroy")
+      : TransactionControl.IMPLICIT;
     try {
-      Transactions.execute(txContext, "destroy", new Callable<Void>() {
+      TransactionContext txContext = context.getTransactionContext();
+      boolean success = Transactions.execute(txContext, "destroy", new Callable<Boolean>() {
```

Comment:
here we do need the extra transaction, because we must call the output committers. See a few lines above.